tests/testConstraintManifold.cpp

```diff
@@ -0,0 +1,195 @@
+  std::cout << "q: " << boost::format("(%0.4f)") % q1 << "\t" << pose_str(pose1)
+            << "\t" << boost::format("(%0.4f)") % q2 << "\t" << pose_str(pose2)
+            << "\n";
+  std::cout << "v: " << boost::format("(%0.4f)") % v1 << "\t" << vec_str(twist1)
+            << "\t" << boost::format("(%0.4f)") % v2 << "\t" << vec_str(twist2)
+            << "\n";
+  std::cout << "a: " << boost::format("(%0.4f)") % a1 << "\t"
+            << vec_str(twista1) << "\t" << boost::format("(%0.4f)") % a2 << "\t"
+            << vec_str(twista2) << "\n";
+  std::cout << "f1: " << vec_str(wrench01) << boost::format("(%0.4f)") % t1
+            << "\t" << vec_str(wrench11) << "\n";
+  std::cout << "f2: " << vec_str(wrench12) << boost::format("(%0.4f)") % t2
+            << "\t" << vec_str(wrench22) << "\n";
+}
+
+TEST(ConstraintManifold, Kinematics) {
+
+  const gtsam::Vector3 gravity(0, 0, -9.8);
+  OptimizerSetting opt;
+  auto graph_builder = DynamicsGraph(opt, gravity);
+
+  // constraints graph
+  NonlinearFactorGraph constraints_graph;
+  constraints_graph.add(graph_builder.dynamicsFactorGraph(robot, 0));
+
+  // initial values
+  Initializer initializer;
+  Values values0 = initializer.ZeroValues(robot, 0, 0.0);
+  Values known_values;
+  for (const auto& joint : robot.joints()) {
+    InsertJointAngle(&known_values, joint->id(),
+                     JointAngle(values0, joint->id()));
+    InsertJointVel(&known_values, joint->id(), JointVel(values0, joint->id()));
+    InsertTorque(&known_values, joint->id(), 0.0);
+  }
+  for (const auto& link : robot.links()) {
+    InsertPose(&known_values, link->id(), Pose(values0, link->id()));
+    InsertTwist(&known_values, link->id(), Twist(values0, link->id()));
+  }
+  values0 = graph_builder.linearSolveFD(robot, 0, known_values);
+  Values init_values = values0;
+
+  // basis keys
+  KeyVector basis_keys;
+  basis_keys.push_back(JointAngleKey(j0_id, 0));
+  basis_keys.push_back(JointAngleKey(j1_id, 0));
+  basis_keys.push_back(JointVelKey(j0_id, 0));
+  basis_keys.push_back(JointVelKey(j1_id, 0));
+  basis_keys.push_back(JointAccelKey(j0_id, 0));
+  basis_keys.push_back(JointAccelKey(j1_id, 0));
+
+  // constraint manifold
+  auto constraints = ConstraintsFromGraph(constraints_graph);
+  auto cc_params = boost::make_shared<ConstraintManifold::Params>();
+  cc_params->retract_type = ConstraintManifold::Params::RetractType::PARTIAL_PROJ;
+  cc_params->basis_type = ConstraintManifold::Params::BasisType::SPECIFY_VARIABLES;
+  auto cc = boost::make_shared<ConnectedComponent>(constraints);
+  auto cm = ConstraintManifold(cc, init_values, cc_params, false, true, basis_keys);
+  print_state(cm.values());
+
+  // retract
+  Vector xi = (Vector(6)<<1, 0, 0, 0, 0, 0).finished();
+  std::cout << "retract\n";
+  auto new_cm = cm.retract(xi);
+  std::cout << "retract done\n";
+  print_state(new_cm.values());
+  std::cout << "constraint violation: " << cc->merit_graph.error(new_cm.values()) << "\n";
```

Comment:
Can you add some assertions?  E.g. assertions for some of the values that get printed out?

Reply:
Temporarily removing the test and replaced with TODO. Will restructure and add back later.